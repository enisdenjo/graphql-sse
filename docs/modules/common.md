[graphql-sse](../README.md) / common

# Module: common

## Table of contents

### Interfaces

- [ExecutionPatchResult](../interfaces/common.ExecutionPatchResult.md)
- [ExecutionResult](../interfaces/common.ExecutionResult.md)
- [RequestParams](../interfaces/common.RequestParams.md)
- [Sink](../interfaces/common.Sink.md)
- [StreamMessage](../interfaces/common.StreamMessage.md)

### Type Aliases

- [StreamData](common.md#streamdata)
- [StreamDataForID](common.md#streamdataforid)
- [StreamEvent](common.md#streamevent)

### Variables

- [TOKEN\_HEADER\_KEY](common.md#token_header_key)
- [TOKEN\_QUERY\_KEY](common.md#token_query_key)

### Functions

- [isAsyncGenerator](common.md#isasyncgenerator)
- [isAsyncIterable](common.md#isasynciterable)
- [parseStreamData](common.md#parsestreamdata)
- [print](common.md#print)
- [validateStreamEvent](common.md#validatestreamevent)

## Common

### StreamData

Ƭ **StreamData**<`E`\>: `E` extends ``"next"`` ? [`ExecutionResult`](../interfaces/common.ExecutionResult.md) \| [`ExecutionPatchResult`](../interfaces/common.ExecutionPatchResult.md) : `E` extends ``"complete"`` ? ``null`` : `never`

#### Type parameters

| Name | Type |
| :------ | :------ |
| `E` | extends [`StreamEvent`](common.md#streamevent) |

___

### StreamDataForID

Ƭ **StreamDataForID**<`E`\>: `E` extends ``"next"`` ? { `id`: `string` ; `payload`: [`ExecutionResult`](../interfaces/common.ExecutionResult.md) \| [`ExecutionPatchResult`](../interfaces/common.ExecutionPatchResult.md)  } : `E` extends ``"complete"`` ? { `id`: `string`  } : `never`

#### Type parameters

| Name | Type |
| :------ | :------ |
| `E` | extends [`StreamEvent`](common.md#streamevent) |

___

### StreamEvent

Ƭ **StreamEvent**: ``"next"`` \| ``"complete"``

___

### TOKEN\_HEADER\_KEY

• `Const` **TOKEN\_HEADER\_KEY**: ``"x-graphql-event-stream-token"``

Header key through which the event stream token is transmitted
when using the client in "single connection mode".

Read more: https://github.com/enisdenjo/graphql-sse/blob/master/PROTOCOL.md#single-connection-mode

___

### TOKEN\_QUERY\_KEY

• `Const` **TOKEN\_QUERY\_KEY**: ``"token"``

URL query parameter key through which the event stream token is transmitted
when using the client in "single connection mode".

Read more: https://github.com/enisdenjo/graphql-sse/blob/master/PROTOCOL.md#single-connection-mode

___

### isAsyncGenerator

▸ **isAsyncGenerator**<`T`\>(`val`): val is AsyncGenerator<T, any, unknown\>

Checkes whether the provided value is an async generator.

#### Type parameters

| Name |
| :------ |
| `T` |

#### Parameters

| Name | Type |
| :------ | :------ |
| `val` | `unknown` |

#### Returns

val is AsyncGenerator<T, any, unknown\>

___

### isAsyncIterable

▸ **isAsyncIterable**<`T`\>(`val`): val is AsyncIterable<T\>

Checkes whether the provided value is an async iterable.

#### Type parameters

| Name |
| :------ |
| `T` |

#### Parameters

| Name | Type |
| :------ | :------ |
| `val` | `unknown` |

#### Returns

val is AsyncIterable<T\>

___

### parseStreamData

▸ **parseStreamData**<`ForID`, `E`\>(`e`, `data`): `ForID` extends ``true`` ? [`StreamDataForID`](common.md#streamdataforid)<`E`\> : [`StreamData`](common.md#streamdata)<`E`\>

#### Type parameters

| Name | Type |
| :------ | :------ |
| `ForID` | extends `boolean` |
| `E` | extends [`StreamEvent`](common.md#streamevent) |

#### Parameters

| Name | Type |
| :------ | :------ |
| `e` | `E` |
| `data` | `string` |

#### Returns

`ForID` extends ``true`` ? [`StreamDataForID`](common.md#streamdataforid)<`E`\> : [`StreamData`](common.md#streamdata)<`E`\>

___

### print

▸ **print**<`ForID`, `E`\>(`msg`): `string`

#### Type parameters

| Name | Type |
| :------ | :------ |
| `ForID` | extends `boolean` |
| `E` | extends [`StreamEvent`](common.md#streamevent) |

#### Parameters

| Name | Type |
| :------ | :------ |
| `msg` | [`StreamMessage`](../interfaces/common.StreamMessage.md)<`ForID`, `E`\> |

#### Returns

`string`

___

### validateStreamEvent

▸ **validateStreamEvent**(`e`): [`StreamEvent`](common.md#streamevent)

#### Parameters

| Name | Type |
| :------ | :------ |
| `e` | `unknown` |

#### Returns

[`StreamEvent`](common.md#streamevent)
