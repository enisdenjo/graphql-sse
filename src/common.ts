/**
 *
 * common
 *
 */

import type { GraphQLError } from 'graphql';
import { isObject } from './utils';

/**
 * Header key through which the event stream token is transmitted
 * when using the client in "single connection mode".
 *
 * Read more: https://github.com/enisdenjo/graphql-sse/blob/master/PROTOCOL.md#single-connection-mode
 *
 * @category Common
 */
export const TOKEN_HEADER_KEY = 'x-graphql-event-stream-token' as const;

/**
 * URL query parameter key through which the event stream token is transmitted
 * when using the client in "single connection mode".
 *
 * Read more: https://github.com/enisdenjo/graphql-sse/blob/master/PROTOCOL.md#single-connection-mode
 *
 * @category Common
 */
export const TOKEN_QUERY_KEY = 'token' as const;

/**
 * Parameters for GraphQL's request for execution.
 *
 * Reference: https://github.com/graphql/graphql-over-http/blob/main/spec/GraphQLOverHTTP.md#request
 *
 * @category Common
 */
export interface RequestParams {
  operationName?: string;
  query: string;
  variables?: Record<string, unknown>;
  extensions?: Record<string, unknown>;
}

/**
 * Represents a message in an event stream.
 *
 * Read more: https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events#Event_stream_format
 *
 * @category Common
 */
export interface StreamMessage<ForID extends boolean, E extends StreamEvent> {
  // id?: string; might be used in future releases for connection recovery
  event: E;
  data: ForID extends true ? StreamDataForID<E> : StreamData<E>;
  // retry?: number; ignored since graphql-sse implements custom retry strategies
}

/** @category Common */
export type StreamEvent = 'next' | 'complete';

/** @category Common */
export function validateStreamEvent(e: unknown): StreamEvent {
  e = e as StreamEvent;
  if (e !== 'next' && e !== 'complete')
    throw new Error(`Invalid stream event "${e}"`);
  return e;
}

/** @category Common */
export function print<ForID extends boolean, E extends StreamEvent>(
  msg: StreamMessage<ForID, E>,
): string {
  let str = `event: ${msg.event}\ndata:`;
  if (msg.data) {
    str += ' ';
    str += JSON.stringify(msg.data);
  }
  str += '\n\n';
  return str;
}

/** @category Common */
export interface ExecutionResult<
  Data = Record<string, unknown>,
  Extensions = Record<string, unknown>,
> {
  errors?: ReadonlyArray<GraphQLError>;
  data?: Data | null;
  hasNext?: boolean;
  extensions?: Extensions;
}

/** @category Common */
export interface ExecutionPatchResult<
  Data = unknown,
  Extensions = Record<string, unknown>,
> {
  errors?: ReadonlyArray<GraphQLError>;
  data?: Data | null;
  path?: ReadonlyArray<string | number>;
  label?: string;
  hasNext: boolean;
  extensions?: Extensions;
}

/** @category Common */
export type StreamData<E extends StreamEvent> = E extends 'next'
  ? ExecutionResult | ExecutionPatchResult
  : E extends 'complete'
  ? null
  : never;

/** @category Common */
export type StreamDataForID<E extends StreamEvent> = E extends 'next'
  ? { id: string; payload: ExecutionResult | ExecutionPatchResult }
  : E extends 'complete'
  ? { id: string }
  : never;

/** @category Common */
export function parseStreamData<ForID extends boolean, E extends StreamEvent>(
  e: E,
  data: string,
) {
  if (data) {
    try {
      data = JSON.parse(data);
    } catch {
      throw new Error('Invalid stream data');
    }
  }

  if (e === 'next' && !data)
    throw new Error('Stream data must be an object for "next" events');

  return (data || null) as ForID extends true
    ? StreamDataForID<E>
    : StreamData<E>;
}

/**
 * A representation of any set of values over any amount of time.
 *
 * @category Common
 */
export interface Sink<T = unknown> {
  /** Next value arriving. */
  next(value: T): void;
  /** An error that has occured. This function "closes" the sink. */
  error(error: unknown): void;
  /** The sink has completed. This function "closes" the sink. */
  complete(): void;
}

/**
 * Checkes whether the provided value is an async iterable.
 *
 * @category Common
 */
export function isAsyncIterable<T>(val: unknown): val is AsyncIterable<T> {
  return typeof Object(val)[Symbol.asyncIterator] === 'function';
}

/**
 * Checkes whether the provided value is an async generator.
 *
 * @category Common
 */
export function isAsyncGenerator<T>(val: unknown): val is AsyncGenerator<T> {
  return (
    isObject(val) &&
    typeof Object(val)[Symbol.asyncIterator] === 'function' &&
    typeof val.return === 'function' &&
    typeof val.throw === 'function' &&
    typeof val.next === 'function'
  );
}
