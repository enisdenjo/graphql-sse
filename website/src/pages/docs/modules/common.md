# Module: common

## Interfaces

- [ExecutionPatchResult](/docs/interfaces/common.ExecutionPatchResult)
- [ExecutionResult](/docs/interfaces/common.ExecutionResult)
- [RequestParams](/docs/interfaces/common.RequestParams)
- [Sink](/docs/interfaces/common.Sink)
- [StreamMessage](/docs/interfaces/common.StreamMessage)

## Common

### StreamData

Ƭ **StreamData**<`E`\>: `E` extends ``"next"`` ? [`ExecutionResult`](/docs/interfaces/common.ExecutionResult) \| [`ExecutionPatchResult`](/docs/interfaces/common.ExecutionPatchResult) : `E` extends ``"complete"`` ? ``null`` : `never`

#### Type parameters

| Name | Type |
| :------ | :------ |
| `E` | extends [`StreamEvent`](/docs/modules/common.md#streamevent) |

___

### StreamDataForID

Ƭ **StreamDataForID**<`E`\>: `E` extends ``"next"`` ? { `id`: `string` ; `payload`: [`ExecutionResult`](/docs/interfaces/common.ExecutionResult) \| [`ExecutionPatchResult`](/docs/interfaces/common.ExecutionPatchResult)  } : `E` extends ``"complete"`` ? { `id`: `string`  } : `never`

#### Type parameters

| Name | Type |
| :------ | :------ |
| `E` | extends [`StreamEvent`](/docs/modules/common.md#streamevent) |

___

### StreamEvent

Ƭ **StreamEvent**: ``"next"`` \| ``"complete"``

___

### TOKEN\_HEADER\_KEY

• `Const` **TOKEN\_HEADER\_KEY**: ``"x-graphql-event-stream-token"``

Header key through which the event stream token is transmitted
when using the client in "single connection mode".

Read more: https://github.com/enisdenjo/graphql-sse/blob/master/PROTOCOL.md#single-connection-mode

___

### TOKEN\_QUERY\_KEY

• `Const` **TOKEN\_QUERY\_KEY**: ``"token"``

URL query parameter key through which the event stream token is transmitted
when using the client in "single connection mode".

Read more: https://github.com/enisdenjo/graphql-sse/blob/master/PROTOCOL.md#single-connection-mode

___

### isAsyncGenerator

▸ **isAsyncGenerator**<`T`\>(`val`): val is AsyncGenerator<T, any, unknown\>

Checkes whether the provided value is an async generator.

#### Type parameters

| Name |
| :------ |
| `T` |

#### Parameters

| Name | Type |
| :------ | :------ |
| `val` | `unknown` |

#### Returns

val is AsyncGenerator<T, any, unknown\>

___

### isAsyncIterable

▸ **isAsyncIterable**<`T`\>(`val`): val is AsyncIterable<T\>

Checkes whether the provided value is an async iterable.

#### Type parameters

| Name |
| :------ |
| `T` |

#### Parameters

| Name | Type |
| :------ | :------ |
| `val` | `unknown` |

#### Returns

val is AsyncIterable<T\>

___

### parseStreamData

▸ **parseStreamData**<`ForID`, `E`\>(`e`, `data`): `ForID` extends ``true`` ? [`StreamDataForID`](/docs/modules/common.md#streamdataforid)<`E`\> : [`StreamData`](/docs/modules/common.md#streamdata)<`E`\>

#### Type parameters

| Name | Type |
| :------ | :------ |
| `ForID` | extends `boolean` |
| `E` | extends [`StreamEvent`](/docs/modules/common.md#streamevent) |

#### Parameters

| Name | Type |
| :------ | :------ |
| `e` | `E` |
| `data` | `string` |

#### Returns

`ForID` extends ``true`` ? [`StreamDataForID`](/docs/modules/common.md#streamdataforid)<`E`\> : [`StreamData`](/docs/modules/common.md#streamdata)<`E`\>

___

### print

▸ **print**<`ForID`, `E`\>(`msg`): `string`

#### Type parameters

| Name | Type |
| :------ | :------ |
| `ForID` | extends `boolean` |
| `E` | extends [`StreamEvent`](/docs/modules/common.md#streamevent) |

#### Parameters

| Name | Type |
| :------ | :------ |
| `msg` | [`StreamMessage`](/docs/interfaces/common.StreamMessage)<`ForID`, `E`\> |

#### Returns

`string`

___

### validateStreamEvent

▸ **validateStreamEvent**(`e`): [`StreamEvent`](/docs/modules/common.md#streamevent)

#### Parameters

| Name | Type |
| :------ | :------ |
| `e` | `unknown` |

#### Returns

[`StreamEvent`](/docs/modules/common.md#streamevent)
