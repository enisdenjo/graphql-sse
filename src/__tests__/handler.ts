import { jest } from '@jest/globals';
import { parse, execute } from 'graphql';
import {
  createTHandler,
  assertString,
  assertAsyncGenerator,
} from './utils/thandler';
import { schema } from './fixtures/simple';
import { TOKEN_HEADER_KEY } from '../common';

it('should only accept valid accept headers', async () => {
  const { handler } = createTHandler();

  let [body, init] = await handler('PUT');
  assertString(body);
  const token = body;

  [body, init] = await handler('GET', {
    headers: {
      accept: 'gibberish',
      [TOKEN_HEADER_KEY]: token,
    },
  });
  expect(init.status).toBe(406);

  [body, init] = await handler('GET', {
    headers: {
      accept: 'application/json',
      [TOKEN_HEADER_KEY]: token,
    },
  });
  expect(init.status).toBe(400);
  expect(init.headers?.['content-type']).toBe(
    'application/json; charset=utf-8',
  );
  expect(body).toMatchInlineSnapshot(
    `"{"errors":[{"message":"Missing query"}]}"`,
  );

  [body, init] = await handler('GET', {
    headers: {
      accept: 'text/event-stream',
    },
  });
  expect(init.status).toBe(400);
  expect(init.headers?.['content-type']).toBe(
    'application/json; charset=utf-8',
  );
  expect(body).toMatchInlineSnapshot(
    `"{"errors":[{"message":"Missing query"}]}"`,
  );
});

it.each(['authenticate', 'onConnect', 'onSubscribe', 'context', 'onOperation'])(
  'should bubble %s errors to the handler',
  async (hook) => {
    const err = new Error('hang hang');
    const { handler } = createTHandler({
      [hook]() {
        throw err;
      },
    });

    await expect(
      handler('POST', {
        headers: {
          accept: 'text/event-stream',
        },
        body: { query: '{ getValue }' },
      }),
    ).rejects.toBe(err);
  },
);

it.each(['onNext', 'onComplete'])(
  'should bubble %s errors to the response body iterator',
  async (hook) => {
    const err = new Error('hang hang');
    const { handler } = createTHandler({
      [hook]() {
        throw err;
      },
    });

    const [stream, init] = await handler('POST', {
      headers: {
        accept: 'text/event-stream',
      },
      body: { query: '{ getValue }' },
    });
    expect(init.status).toBe(200);
    assertAsyncGenerator(stream);

    await expect(
      (async () => {
        for await (const _ of stream) {
          // wait
        }
      })(),
    ).rejects.toBe(err);
  },
);

it('should bubble onNext errors to the response body iterator even if late', async () => {
  const err = new Error('hang hang');
  let i = 0;
  const { handler } = createTHandler({
    onNext() {
      i++;
      if (i > 3) {
        throw err;
      }
    },
  });

  const [stream, init] = await handler('POST', {
    headers: {
      accept: 'text/event-stream',
    },
    body: { query: 'subscription { greetings }' },
  });
  expect(init.status).toBe(200);
  assertAsyncGenerator(stream);

  await expect(
    (async () => {
      for await (const _ of stream) {
        // wait
      }
    })(),
  ).rejects.toBe(err);
});

it('should detect execution args in onSubscribe return value', async () => {
  const executeFn = jest.fn(execute);
  const executionArgs = {
    schema,
    document: parse('{ getValue }'),
    contextValue: undefined,
  };
  const { handler } = createTHandler({
    execute: executeFn,
    onSubscribe() {
      return executionArgs;
    },
  });

  const [stream, init] = await handler('POST', {
    headers: {
      accept: 'text/event-stream',
    },
    body: { query: 'subscription { greetings }' },
  });
  expect(init.status).toBe(200);
  assertAsyncGenerator(stream);

  await stream.next(); // ping
  expect(stream.next()).resolves.toMatchInlineSnapshot(`
    {
      "done": false,
      "value": "event: next
    data: {"data":{"getValue":"value"}}

    ",
    }
  `);

  expect(executeFn).toHaveBeenCalledWith(executionArgs);
});

describe('single connection mode', () => {
  it('should respond with 404s when token was not previously registered', async () => {
    const { handler } = createTHandler();

    let [body, init] = await handler('POST', {
      headers: {
        [TOKEN_HEADER_KEY]: '0',
      },
    });
    expect(init.status).toBe(404);
    expect(init.headers?.['content-type']).toBe(
      'application/json; charset=utf-8',
    );
    expect(body).toMatchInlineSnapshot(
      `"{"errors":[{"message":"Stream not found"}]}"`,
    );

    const search = new URLSearchParams();
    search.set('token', '0');

    [body, init] = await handler('GET', { search });
    expect(init.status).toBe(404);
    expect(init.headers?.['content-type']).toBe(
      'application/json; charset=utf-8',
    );
    expect(body).toMatchInlineSnapshot(
      `"{"errors":[{"message":"Stream not found"}]}"`,
    );

    [body, init] = await handler('DELETE', { search });
    expect(init.status).toBe(404);
    expect(init.headers?.['content-type']).toBe(
      'application/json; charset=utf-8',
    );
    expect(body).toMatchInlineSnapshot(
      `"{"errors":[{"message":"Stream not found"}]}"`,
    );
  });

  it('should get a token with PUT request', async () => {
    const { handler } = createTHandler({
      authenticate() {
        return 'token';
      },
    });

    const [body, init] = await handler('PUT');
    expect(init.status).toBe(201);
    expect(init.headers?.['content-type']).toBe('text/plain; charset=utf-8');
    expect(body).toBe('token');
  });

  it('should treat event streams without reservations as regular requests', async () => {
    const { handler } = createTHandler();

    const [body, init] = await handler('GET', {
      headers: {
        [TOKEN_HEADER_KEY]: '0',
        accept: 'text/event-stream',
      },
    });
    expect(init.status).toBe(400);
    expect(body).toMatchInlineSnapshot(
      `"{"errors":[{"message":"Missing query"}]}"`,
    );
  });

  it('should allow event streams on reservations', async () => {
    const { handler } = createTHandler();

    // token can be sent through headers
    let [token] = await handler('PUT');
    assertString(token);
    let [stream, init] = await handler('GET', {
      headers: {
        [TOKEN_HEADER_KEY]: token,
        accept: 'text/event-stream',
      },
    });
    expect(init.status).toBe(200);
    assertAsyncGenerator(stream);
    stream.return();

    // token can be sent through url search param
    [token] = await handler('PUT');
    assertString(token);
    const search = new URLSearchParams();
    search.set('token', token);
    [stream, init] = await handler('GET', {
      search,
      headers: {
        accept: 'text/event-stream',
      },
    });
    expect(init.status).toBe(200);
    assertAsyncGenerator(stream);
    stream.return();
  });

  it('should not allow operations without providing an operation id', async () => {
    const { handler } = createTHandler();

    const [token] = await handler('PUT');
    assertString(token);

    const [body, init] = await handler('POST', {
      headers: { [TOKEN_HEADER_KEY]: token },
      body: { query: '{ getValue }' },
    });

    expect(init.status).toBe(400);
    expect(body).toMatchInlineSnapshot(
      `"{"errors":[{"message":"Operation ID is missing"}]}"`,
    );
  });

  it('should stream query operations to connected event stream', async () => {
    const { handler } = createTHandler();

    const [token] = await handler('PUT');
    assertString(token);

    const [stream] = await handler('POST', {
      headers: {
        [TOKEN_HEADER_KEY]: token,
        accept: 'text/event-stream',
      },
    });
    assertAsyncGenerator(stream);

    const [body, init] = await handler('POST', {
      headers: { [TOKEN_HEADER_KEY]: token },
      body: { query: '{ getValue }', extensions: { operationId: '1' } },
    });
    expect(init.status).toBe(202);
    expect(body).toBeNull();

    await expect(stream.next()).resolves.toMatchInlineSnapshot(`
      {
        "done": false,
        "value": ":

      ",
      }
    `); // ping

    await expect(stream.next()).resolves.toMatchInlineSnapshot(`
      {
        "done": false,
        "value": "event: next
      data: {"id":"1","payload":{"data":{"getValue":"value"}}}

      ",
      }
    `);

    await expect(stream.next()).resolves.toMatchInlineSnapshot(`
      {
        "done": false,
        "value": "event: complete
      data: {"id":"1"}

      ",
      }
    `);

    stream.return();
  });

  it('should stream subscription operations to connected event stream', async () => {
    const { handler } = createTHandler();

    const [token] = await handler('PUT');
    assertString(token);

    const search = new URLSearchParams();
    search.set('token', token);
    const [stream] = await handler('GET', {
      search,
      headers: {
        accept: 'text/event-stream',
      },
    });
    assertAsyncGenerator(stream);
    await expect(stream.next()).resolves.toMatchInlineSnapshot(`
      {
        "done": false,
        "value": ":

      ",
      }
    `); // ping

    const [_0, init] = await handler('POST', {
      headers: {
        [TOKEN_HEADER_KEY]: token,
      },
      body: {
        query: 'subscription { greetings }',
        extensions: { operationId: '1' },
      },
    });
    expect(init.status).toBe(202);

    for await (const msg of stream) {
      expect(msg).toMatchSnapshot();

      if (msg.startsWith('event: complete')) {
        break;
      }
    }

    await expect(stream.next()).resolves.toMatchInlineSnapshot(`
      {
        "done": true,
        "value": undefined,
      }
    `);
  });

  it.todo('should stream operations even if event stream connects late');

  it('should report validation issues to operation request', async () => {
    const { handler } = createTHandler();

    const [token] = await handler('PUT');
    assertString(token);

    const search = new URLSearchParams();
    search.set('token', token);
    const [stream] = await handler('GET', {
      search,
      headers: {
        accept: 'text/event-stream',
      },
    });
    assertAsyncGenerator(stream);
    await expect(stream.next()).resolves.toMatchInlineSnapshot(`
      {
        "done": false,
        "value": ":

      ",
      }
    `); // ping

    const [body, init] = await handler('POST', {
      headers: {
        [TOKEN_HEADER_KEY]: token,
      },
      body: {
        query: 'subscription { notExists }',
        extensions: { operationId: '1' },
      },
    });
    expect(init.status).toBe(400);
    expect(body).toMatchInlineSnapshot(
      `"{"errors":[{"message":"Cannot query field \\"notExists\\" on type \\"Subscription\\".","locations":[{"line":1,"column":16}]}]}"`,
    );

    // stream remains open
    await expect(
      Promise.race([
        stream.next(),
        await new Promise((resolve) => setTimeout(resolve, 20)),
      ]),
    ).resolves.toBeUndefined();
    stream.return();
  });

  it('should report subscription errors to the stream', async () => {
    const { handler } = createTHandler();

    const [token] = await handler('PUT');
    assertString(token);

    const search = new URLSearchParams();
    search.set('token', token);
    const [stream] = await handler('GET', {
      search,
      headers: {
        accept: 'text/event-stream',
      },
    });
    assertAsyncGenerator(stream);
    await expect(stream.next()).resolves.toMatchInlineSnapshot(`
      {
        "done": false,
        "value": ":

      ",
      }
    `); // ping

    const [, init] = await handler('POST', {
      headers: {
        [TOKEN_HEADER_KEY]: token,
      },
      body: {
        query: 'subscription { throwing }',
        extensions: { operationId: '1' },
      },
    });
    expect(init.status).toBe(202);

    await expect(stream.next()).resolves.toMatchInlineSnapshot(`
      {
        "done": false,
        "value": "event: next
      data: {"id":"1","payload":{"errors":[{"message":"Kaboom!","locations":[{"line":1,"column":16}],"path":["throwing"]}]}}

      ",
      }
    `);

    stream.return();
  });
});

describe('distinct connections mode', () => {
  it('should stream query operations to connected event stream and then disconnect', async () => {
    const { handler } = createTHandler();

    // GET
    const search = new URLSearchParams();
    search.set('query', '{ getValue }');
    let [stream, init] = await handler('GET', {
      search,
      headers: {
        accept: 'text/event-stream',
      },
    });
    expect(init.status).toBe(200);
    assertAsyncGenerator(stream);
    for await (const msg of stream) {
      expect(msg).toMatchSnapshot();
    }

    // POST
    [stream, init] = await handler('POST', {
      headers: {
        accept: 'text/event-stream',
      },
      body: { query: '{ getValue }' },
    });
    expect(init.status).toBe(200);
    assertAsyncGenerator(stream);
    for await (const msg of stream) {
      expect(msg).toMatchSnapshot();
    }
  });

  it('should stream subscription operations to connected event stream and then disconnect', async () => {
    const { handler } = createTHandler();

    // GET
    const search = new URLSearchParams();
    search.set('query', 'subscription { greetings }');
    let [stream, init] = await handler('GET', {
      search,
      headers: {
        accept: 'text/event-stream',
      },
    });
    expect(init.status).toBe(200);
    assertAsyncGenerator(stream);
    for await (const msg of stream) {
      expect(msg).toMatchSnapshot();
    }

    // POST
    [stream, init] = await handler('POST', {
      headers: {
        accept: 'text/event-stream',
      },
      body: { query: 'subscription { greetings }' },
    });
    expect(init.status).toBe(200);
    assertAsyncGenerator(stream);
    for await (const msg of stream) {
      expect(msg).toMatchSnapshot();
    }
  });

  it('should report operation validation issues by streaming them', async () => {
    const { handler } = createTHandler();

    const [stream, init] = await handler('POST', {
      headers: {
        accept: 'text/event-stream',
      },
      body: { query: '{ notExists }' },
    });
    expect(init.status).toBe(200);
    assertAsyncGenerator(stream);
    await expect(stream.next()).resolves.toMatchInlineSnapshot(`
      {
        "done": false,
        "value": ":

      ",
      }
    `); // ping
    await expect(stream.next()).resolves.toMatchInlineSnapshot(`
      {
        "done": false,
        "value": "event: next
      data: {"errors":[{"message":"Cannot query field \\"notExists\\" on type \\"Query\\".","locations":[{"line":1,"column":3}]}]}

      ",
      }
    `);
    await expect(stream.next()).resolves.toMatchInlineSnapshot(`
      {
        "done": false,
        "value": "event: complete
      data:

      ",
      }
    `);
    await expect(stream.next()).resolves.toMatchInlineSnapshot(`
      {
        "done": true,
        "value": undefined,
      }
    `);
  });

  it('should complete subscription operations after client disconnects', async () => {
    const { handler } = createTHandler();

    const [stream, init] = await handler('POST', {
      headers: {
        accept: 'text/event-stream',
      },
      body: { query: `subscription { ping(key: "${Math.random()}") }` },
    });
    expect(init.status).toBe(200);
    assertAsyncGenerator(stream);

    // simulate client disconnect in next tick
    setTimeout(() => stream.return(), 0);

    for await (const _ of stream) {
      // loop must break for test to pass
    }
  });

  it('should complete when stream ends before the subscription sent all events', async () => {
    const { handler } = createTHandler();

    const [stream, init] = await handler('POST', {
      headers: {
        accept: 'text/event-stream',
      },
      body: { query: `subscription { greetings }` },
    });
    expect(init.status).toBe(200);
    assertAsyncGenerator(stream);

    await expect(stream.next()).resolves.toMatchInlineSnapshot(`
      {
        "done": false,
        "value": ":

      ",
      }
    `); // ping

    for await (const msg of stream) {
      expect(msg).toMatchInlineSnapshot(`
        "event: next
        data: {"data":{"greetings":"Hi"}}

        "
      `);

      // return after first message (there are more)
      break;
    }

    // message was already queued up (pending), it's ok to have it
    await expect(stream.next()).resolves.toMatchInlineSnapshot(`
      {
        "done": false,
        "value": "event: next
      data: {"data":{"greetings":"Bonjour"}}

      ",
      }
    `);

    await expect(stream.next()).resolves.toMatchInlineSnapshot(`
      {
        "done": false,
        "value": "event: complete
      data:

      ",
      }
    `);
    await expect(stream.next()).resolves.toMatchInlineSnapshot(`
      {
        "done": true,
        "value": undefined,
      }
    `);
  });
});
