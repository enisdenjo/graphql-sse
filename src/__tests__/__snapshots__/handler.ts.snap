// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`distinct connections mode should stream query operations to connected event stream and then disconnect 1`] = `
":

"
`;

exports[`distinct connections mode should stream query operations to connected event stream and then disconnect 2`] = `
"event: next
data: {\\"data\\":{\\"getValue\\":\\"value\\"}}

"
`;

exports[`distinct connections mode should stream query operations to connected event stream and then disconnect 3`] = `
"event: complete

"
`;

exports[`distinct connections mode should stream query operations to connected event stream and then disconnect 4`] = `
":

"
`;

exports[`distinct connections mode should stream query operations to connected event stream and then disconnect 5`] = `
"event: next
data: {\\"data\\":{\\"getValue\\":\\"value\\"}}

"
`;

exports[`distinct connections mode should stream query operations to connected event stream and then disconnect 6`] = `
"event: complete

"
`;

exports[`distinct connections mode should stream subscription operations to connected event stream and then disconnect 1`] = `
":

"
`;

exports[`distinct connections mode should stream subscription operations to connected event stream and then disconnect 2`] = `
"event: next
data: {\\"data\\":{\\"greetings\\":\\"Hi\\"}}

"
`;

exports[`distinct connections mode should stream subscription operations to connected event stream and then disconnect 3`] = `
"event: next
data: {\\"data\\":{\\"greetings\\":\\"Bonjour\\"}}

"
`;

exports[`distinct connections mode should stream subscription operations to connected event stream and then disconnect 4`] = `
"event: next
data: {\\"data\\":{\\"greetings\\":\\"Hola\\"}}

"
`;

exports[`distinct connections mode should stream subscription operations to connected event stream and then disconnect 5`] = `
"event: next
data: {\\"data\\":{\\"greetings\\":\\"Ciao\\"}}

"
`;

exports[`distinct connections mode should stream subscription operations to connected event stream and then disconnect 6`] = `
"event: next
data: {\\"data\\":{\\"greetings\\":\\"Zdravo\\"}}

"
`;

exports[`distinct connections mode should stream subscription operations to connected event stream and then disconnect 7`] = `
"event: complete

"
`;

exports[`distinct connections mode should stream subscription operations to connected event stream and then disconnect 8`] = `
":

"
`;

exports[`distinct connections mode should stream subscription operations to connected event stream and then disconnect 9`] = `
"event: next
data: {\\"data\\":{\\"greetings\\":\\"Hi\\"}}

"
`;

exports[`distinct connections mode should stream subscription operations to connected event stream and then disconnect 10`] = `
"event: next
data: {\\"data\\":{\\"greetings\\":\\"Bonjour\\"}}

"
`;

exports[`distinct connections mode should stream subscription operations to connected event stream and then disconnect 11`] = `
"event: next
data: {\\"data\\":{\\"greetings\\":\\"Hola\\"}}

"
`;

exports[`distinct connections mode should stream subscription operations to connected event stream and then disconnect 12`] = `
"event: next
data: {\\"data\\":{\\"greetings\\":\\"Ciao\\"}}

"
`;

exports[`distinct connections mode should stream subscription operations to connected event stream and then disconnect 13`] = `
"event: next
data: {\\"data\\":{\\"greetings\\":\\"Zdravo\\"}}

"
`;

exports[`distinct connections mode should stream subscription operations to connected event stream and then disconnect 14`] = `
"event: complete

"
`;
